pallets/handles/src/lib.rs

```diff
@@ -0,0 +1,628 @@
+		/// * `payload` - The payload containing the base handle to claim.
+		///
+		/// # Returns
+		///
+		/// * `Handle` - The full display handle.
+		///
+		pub fn do_claim_handle(
+			delegator_msa_id: MessageSourceId,
+			payload: ClaimHandlePayload,
+		) -> Result<Vec<u8>, DispatchError> {
+			// Validation: The MSA must not already have a handle associated with it
+			ensure!(
+				MSAIdToDisplayName::<T>::try_get(delegator_msa_id).is_err(),
+				Error::<T>::MSAHandleAlreadyExists
+			);
+
+			// Convert base handle to UTF-8 string slice while validating.
+			let base_handle_str = core::str::from_utf8(&payload.base_handle)
+				.map_err(|_| Error::<T>::InvalidHandleEncoding)?;
+
+			// Validation: The handle length must be valid.
+			// Note: the count() can panic but won't because the base_handle byte length is already checked
+			let len = base_handle_str.chars().count() as u32;
+
+			// Validation: Handle character length must be within range
+			ensure!(
+				len >= HANDLE_BASE_CHARS_MIN && len <= HANDLE_BASE_CHARS_MAX,
+				Error::<T>::InvalidHandleCharacterLength
+			);
+
+			// Validation: The handle must consist of characters not containing reserved words or blocked characters
+			let handle_validator = HandleValidator::new();
+			ensure!(
+				handle_validator.consists_of_supported_unicode_character_sets(base_handle_str),
+				Error::<T>::HandleDoesNotConsistOfSupportedCharacterSets
+			);
+			ensure!(
+				!handle_validator.is_reserved_handle(base_handle_str),
+				Error::<T>::HandleIsNotAllowed
+			);
+			ensure!(
+				!handle_validator.contains_blocked_characters(base_handle_str),
+				Error::<T>::HandleContainsBlockedCharacters
+			);
+
+			// Convert base display handle into a canonical display handle
+			let canonical_handle_str = HandleConverter::convert_to_canonical(&base_handle_str);
+			let canonical_handle_vec = canonical_handle_str.as_bytes().to_vec();
+			let canonical_handle: Handle = canonical_handle_vec.try_into().unwrap();
+
+			// Generate suffix from the next available index into the suffix sequence
+			let suffix_sequence_index =
+				Self::get_next_suffix_index_for_canonical_handle(canonical_handle.clone())?;
+			let suffix = Self::generate_suffix_for_canonical_handle(
+				&canonical_handle_str,
+				suffix_sequence_index as usize,
+			);
+
+			// Store canonical handle and suffix to MSA id
+			CanonicalBaseHandleAndSuffixToMSAId::<T>::insert(
+				canonical_handle.clone(),
+				suffix,
+				delegator_msa_id,
+			);
+			// Store canonical handle to suffix sequence index
+			CanonicalBaseHandleToSuffixIndex::<T>::set(
+				canonical_handle.clone(),
+				Some(suffix_sequence_index),
+			);
+
+			// Compose the full display handle from the base handle, "." delimiter and suffix
+			let mut full_handle_vec: Vec<u8> = vec![];
+			full_handle_vec.extend(base_handle_str.as_bytes());
+			full_handle_vec.extend(b"."); // The delimiter
+			let mut buff = [0u8; SUFFIX_MAX_DIGITS];
```

Comment:
SUFFIX_MAX_DIGITS maybe this also belong in the config?

Reply:
I think this is going to keep us afloat for years, hence we kept it as constant

Reply:
Moved to https://github.com/LibertyDSNP/frequency/issues/1398